openlibrary/core/observations.py

```diff
@@ -1,26 +1,252 @@
+    """
+    observations = Observations.get_observation_types_and_values()
+
+    observation_dict = {}
+
+    for o in observations:
+        type = o['type']
+
+        if type not in observation_dict:
+            observation_dict[type] = {
+                'id': o['type_id'],
+                'label': type,
+                'description': o['description'],
+                'multi_choice': o['multi_choice'],
+                'values': []
+            }
+
+        observation_dict[type]['values'].insert(0, ObservationValue(o['value_id'], o['value'], o['prev_value']))
+
+    observation_list = []
 
-# URL for TheBestBookOn
-TBBO_URL = config.get('tbbo_url')
+    for v in observation_dict.values():
+        v['values'] = _sort_values(v['values'])
+        # Do not include observation type if it has no values:
+        if len(v['values']):
+            observation_list.append(v)
 
-def post_observation(data, s3_keys):
-    headers = {
-        'x-s3-access': s3_keys['access'],
-        'x-s3-secret': s3_keys['secret']
+    response = {
+        'observations': observation_list
     }
 
-    response = requests.post(TBBO_URL + '/api/observations', data=data, headers=headers)
+    return response
+
+def _sort_values(values_list):
+    """
+    Given a list of one or more ObservationValue tuples, returns a sorted list of observation values.
+
+    ObservationValues are namedtuples with fields ('value_id', 'value', 'prev_value_id').
+    Each field maps to a column in the observation_values table:
+
+    'value_id' -> observation_values.id
+    'value' -> observations_values.value
+    'prev_value_id' -> observation_values.prev_value
+
+    Before values are displayed to patrons in a form, they must first be ordered.  Order is
+    maintained by the observation_values.prev_value column.  Each row has a reference to the
+    ID of the preceding row, or None if the row is the first value in the list.
+
+    return: A sorted list of values.
+    """
+
+    if not len(values_list):
+        return []
+
+    # Add middle list item to sorted list
+    middle_item = values_list.pop(len(values_list) // 2)
+    sorted_list = [ middle_item.value ]
+
+    # Previous id:
+    prev = middle_item.prev_value_id
+
+    # Next id:
+    next = middle_item.value_id
+
+    unsorted_values_dict = { i.value_id: i for i in values_list }
+
+    while len(unsorted_values_dict):
+        for key in list(unsorted_values_dict):   # unsorted_values_dict will change during iteration
+            value = unsorted_values_dict
+            if value[key].value_id == prev:
+                sorted_list.insert(0, value[key].value)
+                prev = value[key].prev_value_id
+                del unsorted_values_dict[key]
+            elif value[key].prev_value_id == next:
+                sorted_list.append(value[key].value)
+                next = value[key].value_id
+                del unsorted_values_dict[key]
```

Comment:
```suggestion
            item = unsorted_values_dict[key]
            if item.value_id == prev:
                sorted_list.insert(0, item.value)
                prev = item.prev_value_id
                del unsorted_values_dict[key]
            elif item.prev_value_id == next:
                sorted_list.append(item.value)
                next = item.value_id
                del unsorted_values_dict[key]
```
Normally dicts have keys and values but that is `value` in two ways in this code so let's use `item` instead.